cmd/clusterctl/client/cluster/mover.go

```diff
@@ -592,6 +805,103 @@ func (o *objectMover) createTargetObject(nodeToCreate *node, toProxy Proxy) erro
 	return nil
 }
 
+func (o *objectMover) saveTargetObject(nodeToCreate *node, directory string) error {
+	log := logf.Log
+	log.V(1).Info("Saving", nodeToCreate.identity.Kind, nodeToCreate.identity.Name, "Namespace", nodeToCreate.identity.Namespace)
+
+	cFrom, err := o.fromProxy.NewClient()
+	if err != nil {
+		return err
+	}
+
+	// Get the source object
+	obj := &unstructured.Unstructured{}
+	obj.SetAPIVersion(nodeToCreate.identity.APIVersion)
+	obj.SetKind(nodeToCreate.identity.Kind)
+	objKey := client.ObjectKey{
+		Namespace: nodeToCreate.identity.Namespace,
+		Name:      nodeToCreate.identity.Name,
+	}
+
+	if err := cFrom.Get(ctx, objKey, obj); err != nil {
+		return errors.Wrapf(err, "error reading %q %s/%s",
+			obj.GroupVersionKind(), obj.GetNamespace(), obj.GetName())
+	}
+
+	// Get JSON for object and write it into the configured directory
+	byObj, err := obj.MarshalJSON()
+	if err != nil {
+		return err
+	}
+
+	filenameObj := nodeToCreate.identity.Kind + "_" + nodeToCreate.identity.Name + "_" + nodeToCreate.identity.Namespace
+	objectFile := filepath.Join(directory, filenameObj)
+
+	err = ioutil.WriteFile(objectFile, byObj, 0644)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *objectMover) restoreTargetObject(nodeToCreate *node, toProxy Proxy) error {
```

Comment:
I suggest to change a little bit this func to make it possible to restart a restore after an error.
In order to do so we should try to get the object as a first thing; if it already exists, then it is required only to update the UID in order to make it possible o rebuild the owner chain.

Reply:
Let's have a unit test for this func ensuring:
- It creates objects as expected
- It does not overrides existing objects (but it fetches the the UID)

Reply:
Unit test are in place, but IMO we still need to improve the restoreTargetObject func.
My idea here here is that what is already in the cluster should not be touched by restore (because what is in the cluster is more recent that what is the files)
Accordingly, if get returns an object, we should only update the node inside the object graph with current object UID, so the dependants owner chain get properly recreated, but we should not touch the object itself in the cluster.
Does this seems reasonable to you?

Reply:
I see - I think that helps clarify. So instead of doing `Update` if the node already exists, we simply continue since we've updated the reference chain. I think all that would need to be changed is [this part here](https://github.com/kubernetes-sigs/cluster-api/pull/4808/files#diff-3458bf9e9c245fe801573500c33de535f05d3a5887f080a405b819e3c0fed415R922-R925) since before this, we attempt a `Get` and update the objects UID based on any existing object. Then, we rebuild the owner references with the UID between L:883-903. We then attempt a `Create` by default and instead of it updating if the object already exists, simply continue (maybe log something out).

Reply:
I think you got the idea right. With regard to the implementation I struggle I little bit to follow what you are suggesting (the link above does not work for me 😞 ), but hopefully this can help.
- Get the object corresponding to the node to create
- If the object already exists, 
   - log a message
   - update `nodeToCreate.newUID=existingTargetObj.GetUID()` so the owner chain will be properly rebuilt 
  - return
- if the object does not exists, create it (in this case apierrors.IsAlreadyExists(err) should never happen, so you can cleanup a good chunk of the existing logic)

Note. This should be covered by a test

Reply:
Let me know what you think of my implementation for this! Happy to refactor :smile: